frame/assets/src/functions.rs

```diff
@@ -45,34 +61,41 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	pub(super) fn new_account(
 		who: &T::AccountId,
 		d: &mut AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>,
-	) -> Result<bool, DispatchError> {
+		maybe_deposit: Option<DepositBalanceOf<T, I>>,
+	) -> Result<ExistenceReason<DepositBalanceOf<T, I>>, DispatchError> {
 		let accounts = d.accounts.checked_add(1).ok_or(ArithmeticError::Overflow)?;
-		let is_sufficient = if d.is_sufficient {
+		let reason = if let Some(deposit) = maybe_deposit {
+			ExistenceReason::DepositHeld(deposit)
+		} else if d.is_sufficient {
 			frame_system::Pallet::<T>::inc_sufficients(who);
 			d.sufficients += 1;
-			true
+			ExistenceReason::Sufficient
 		} else {
 			frame_system::Pallet::<T>::inc_consumers(who).map_err(|_| Error::<T, I>::NoProvider)?;
-			false
+			ExistenceReason::Consumer
 		};
 		d.accounts = accounts;
-		Ok(is_sufficient)
+		Ok(reason)
 	}
 
 	pub(super) fn dead_account(
 		what: T::AssetId,
 		who: &T::AccountId,
 		d: &mut AssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T, I>>,
-		sufficient: bool,
-	) {
-		if sufficient {
-			d.sufficients = d.sufficients.saturating_sub(1);
-			frame_system::Pallet::<T>::dec_sufficients(who);
-		} else {
-			frame_system::Pallet::<T>::dec_consumers(who);
+		reason: &ExistenceReason<DepositBalanceOf<T, I>>,
+	) -> DeadConsequence {
+		match *reason {
+			ExistenceReason::Consumer => frame_system::Pallet::<T>::dec_consumers(who),
+			ExistenceReason::Sufficient => {
+				d.sufficients = d.sufficients.saturating_sub(1);
+				frame_system::Pallet::<T>::dec_sufficients(who);
+			},
+			ExistenceReason::DepositHeld(_) => return Keep,
+			ExistenceReason::DepositRefunded => {},
 		}
 		d.accounts = d.accounts.saturating_sub(1);
-		T::Freezer::died(what, who)
+		T::Freezer::died(what, who);
```

Comment:
sidenote: I'm not sure this is very good to execute some callback while some change in storage are still pending.
Here we have details in a mutable variable, if the `died` callback somehow do some creation on the asset then when committing the details later, all the changes will be overriden.

Reply:
I agree this can be worrying, but we should just document what is allowable in this callback function, for example, that it should not inspect any storage within the assets pallet while making this callback.

Reply:
yes, but if it is possible to not have the constraint it would be better, the doc will be easily forgot. and it makes the API difficult

Reply:
This PR doesn't change anything in this regard.

Reply:
yes indeed, also here is the issue https://github.com/paritytech/substrate/issues/10432